core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala

```diff
@@ -0,0 +1,236 @@
+          groupId = groupId,
+          memberId = "non-exist",
+          generationId = memberEpoch,
+          offset = 200 + version,
+          version = version.toShort,
+          expectedTxnCommitError = Errors.UNKNOWN_MEMBER_ID
+        )
+
+        // Verify TXN_OFFSET_COMMIT request failed with incorrect generationId
+        verifyTxnCommitAndFetch(
+          topic = topic,
+          partition = partition,
+          transactionalId = transactionalId,
+          groupId = groupId,
+          memberId = memberId,
+          generationId = 100,
+          offset = 200 + version,
+          version = version.toShort,
+          expectedTxnCommitError = Errors.ILLEGAL_GENERATION
+        )
+      } else {
+        // Verify that the TXN_OFFSET_COMMIT request failed when group metadata is set under version 3.
+        assertThrows(classOf[UnsupportedVersionException], () =>
+          verifyTxnCommitAndFetch(
+            topic = topic,
+            partition = partition,
+            transactionalId = transactionalId,
+            groupId = groupId,
+            memberId = memberId,
+            generationId = memberEpoch,
+            offset = 200 + version,
+            version = version.toShort,
+            expectedTxnCommitError = Errors.NONE
+          )
+        )
+      }
+    }
+  }
+
+  private def verifyTxnCommitAndFetch(
+    topic: String,
+    partition: Int,
+    transactionalId: String,
+    groupId: String,
+    memberId: String,
+    generationId: Int,
+    offset: Long,
+    version: Short,
+    expectedTxnCommitError: Errors
+  ): Unit = {
+    var producerIdAndEpoch: ProducerIdAndEpoch = null
+    // Wait until ALLOCATE_PRODUCER_ID request finished
+    TestUtils.waitUntilTrue(() =>
+      try {
+        producerIdAndEpoch = initProducerId(
+          transactionalId = transactionalId,
+          producerIdAndEpoch = ProducerIdAndEpoch.NONE,
+          expectedError = Errors.NONE
+        )
+        true
+      } catch {
+        case _: Throwable => false
+      }, "initProducerId request failed"
+    )
+
+    addOffsetsToTxn(
+      groupId = groupId,
+      producerId = producerIdAndEpoch.producerId,
+      producerEpoch = producerIdAndEpoch.epoch,
+      transactionalId = transactionalId
+    )
+
+    commitTxnOffset(
+      groupId = groupId,
+      memberId = memberId,
+      generationId = generationId,
+      producerId = producerIdAndEpoch.producerId,
+      producerEpoch = producerIdAndEpoch.epoch,
+      transactionalId = transactionalId,
+      topic = topic,
+      partition = partition,
+      offset = offset,
+      expectedError = expectedTxnCommitError,
+      version = version
+    )
+
+    endTxn(
+      producerId = producerIdAndEpoch.producerId,
+      producerEpoch = producerIdAndEpoch.epoch,
+      transactionalId = transactionalId,
+      isTransactionV2Enabled = false,
+      committed = true,
+      expectedError = Errors.NONE
+    )
+
+    if (expectedTxnCommitError != Errors.NONE) return
```

Comment:
Could you please confirm that committed offsets remain unchanged after a failed `TXN_OFFSET_COMMIT`?

Reply:
Fixed in the latest commit, thanks.